File: Source/CSharp/Expressive/Expressive/Functions/SignFunction.cs
using Expressive.Expressions;
using System;

namespace Expressive.Functions
{
    internal class SignFunction : FunctionBase
    {
        #region FunctionBase Members

        public override string Name { get { return "Sign"; } }

        public override object Evaluate(IExpression[] parameters)
        {
            this.ValidateParameterCount(parameters, 1, 1);

            var value = parameters[0].Evaluate(Variables);

            if (value != null)
            {
                var valueType = Type.GetTypeCode(value.GetType());

                switch (valueType)
                {
                    case TypeCode.Decimal:
                        return Math.Sign(Convert.ToDecimal(value));
                    case TypeCode.Double:
                        return Math.Sign(Convert.ToDouble(value));
                    case TypeCode.Int16:
                        return Math.Sign(Convert.ToInt16(value));
                    case TypeCode.UInt16:
                        return Math.Sign(Convert.ToUInt16(value));
                    case TypeCode.Int32:
                        return Math.Sign(Convert.ToInt32(value));
                    case TypeCode.UInt32:
                        return Math.Sign(Convert.ToUInt32(value));
                    case TypeCode.Int64:
                        return Math.Sign(Convert.ToInt64(value));
                    case TypeCode.SByte:
                        return Math.Sign(Convert.ToSByte(value));
                    case TypeCode.Single:
                        return Math.Sign(Convert.ToSingle(value));
                    default:
                        break;
                }
            }

            return null;
        }

        #endregion
    }
}

File: Source/Java/Expressive/src/main/java/com/bijington/expressive/functions/mathematical/SignFunction.java
package com.bijington.expressive.functions.mathematical;

import com.bijington.expressive.exceptions.ExpressiveException;
import com.bijington.expressive.expressions.IExpression;
import com.bijington.expressive.functions.FunctionBase;
import com.bijington.expressive.helpers.Convert;

/**
 * Created by shaun on 20/12/2016.
 */
public class SignFunction extends FunctionBase {
    @Override
    public String getName() {
        return "Sign";
    }

    @Override
    public Object evaluate(IExpression[] parameters) throws ExpressiveException {
        this.validateParameterCount(parameters, 1, 1);

        Object value = parameters[0].evaluate(this.getVariables());

        if (value == null) return null;

        return Convert.toDouble(value).compareTo(0.0);
    }
}
